worker/src/main/java/gov/cms/ab2d/worker/adapter/bluebutton/ContractAdapterImpl.java

```diff
@@ -0,0 +1,186 @@
+import org.hl7.fhir.dstu3.model.ResourceType;
+import org.springframework.stereotype.Component;
+
+import java.text.ParseException;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+
+@Slf4j
+//@Primary  - once the BFD API starts returning data, change this to primary bean so spring injects this instead of the stub.
+@Component
+@RequiredArgsConstructor
+public class ContractAdapterImpl implements ContractAdapter {
+
+    private final BFDClient bfdClient;
+
+
+
+    @Override
+    public GetPatientsByContractResponse getPatients(String contractNumber) {
+
+        var currentMonth = LocalDate.now().getMonthValue();
+        log.info("Current Month : [{}]", currentMonth);
+
+        var patientDTOs = new ArrayList<PatientDTO>();
+
+        for (var month = 1; month <= currentMonth; month++) {
+            var bfdPatientsIds = getPatientIdsForMonth(contractNumber, month);
+
+            for (String bfdPatientId : bfdPatientsIds) {
+                var activeDateRange = toDateRange(month);
+
+                var optPatient = findPatient(patientDTOs, bfdPatientId);
+                if (optPatient.isPresent()) {
+                    // patient id was already active on this contract in previous month(s)
+                    // So just add this month to the patient's datesUnderContract
+
+                    var patientDTO = optPatient.get();
+                    if (activeDateRange != null) {
+                        patientDTO.getDatesUnderContract().add(activeDateRange);
+                    }
+
+                } else {
+                    // new patient id.
+                    // Create a new PatientDTO for this patient
+                    // And then add this month to the patient's datesUnderContract
+
+                    var patientDTO = PatientDTO.builder()
+                            .patientId(bfdPatientId)
+                            .build();
+
+                    if (activeDateRange != null) {
+                        patientDTO.getDatesUnderContract().add(activeDateRange);
+                    }
+
+                    patientDTOs.add(patientDTO);
+                }
+            }
+        }
+
+        return GetPatientsByContractResponse.builder()
+                .contractNumber(contractNumber)
+                .patients(patientDTOs)
+                .build();
+    }
+
+
+    /**
+     * Given a contractNumber and month,
+     * calls BFD api to get a bundle of resources from which the patientIDs are extracted.
+     * If there are multiple pages available, fetches all pages.
+     *
+     * @param contractNumber
+     * @param month
+     * @return a list of PatientIds
+     */
+    private List<String> getPatientIdsForMonth(String contractNumber, Integer month) {
+        Bundle bundle = getBundle(contractNumber, month);
+        final List<String> patientIDs = extractPatientIDs(bundle);
```

Comment:
Should we use a `Set` to cover a scenario (albeit unlikely) BFD returns duplicates for a _given_ month?

Reply:
Good point.  Will change it to use `Set` to avoid duplicates.

Reply:
done